File: 53-MaximumSubarray/53-MaximumSubarray.cpp
// Last updated: 9/14/2025, 9:18:11 AM
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int sum = 0;
        int maxSum = INT_MIN;
        for(int i=0;i<nums.size();i++){
            sum += nums[i];
            maxSum = std::max(sum,maxSum);
            if(sum < 0){
                sum = 0;
            }
        }

        return maxSum;
    }
};

File: 53-MaximumSubarray/53-MaximumSubarray.java
// Last updated: 9/14/2025, 9:18:07 AM
class Solution {
    public int maxSubArray(int[] nums) {
        //Optimal solution (Kadane's algo)

        int maxSum = Integer.MIN_VALUE;
        int sum = 0;

        for(int i:nums){
            sum+=i;
            maxSum = Integer.max(maxSum,sum);
            if(sum<0) sum = 0;
        }

        return maxSum;
    }
}